mmdet3d/models/roi_heads/roi_extractors/bev_feature_extractor.py

```diff
@@ -0,0 +1,148 @@
+        """Forward function to BEV feature extractor.
+
+        Args:
+            bev_feature (list[torch.Tensor]): Multi-level feature maps. The
+                 shape of each feature map is [B, C_i, H_i, W_i]
+            rois (list[list[bboxes, ...]]): Decoded bbox, scores and labels
+                The out list indicates the rois in a batch.
+                - bboxes (:obj:`BaseInstance3DBoxes`): Prediction bboxes
+
+        Returns:
+            list[torch.Tensor]: roi features with shape of [N, 5*C]
+        """
+
+        # NOTE: ONLY surpport the latest output of BEV feature maps now
+        bev_feature = bev_feature[-1]
+
+        roi_features = []
+        batch_size = len(bev_feature)
+        num_points = 5
+        for i in range(batch_size):
+            bboxes = rois[i][0]
+            num_boxes = len(bboxes)
+            feature_points = self.get_feature_points(bboxes)  # [5*num_box, 2]
+
+            xs = (feature_points[..., 0] - self.pc_start[0]
+                  ) / self.voxel_size[0] / self.downsample_stride
+            ys = (feature_points[..., 1] - self.pc_start[1]
+                  ) / self.voxel_size[1] / self.downsample_stride
+
+            features = self.bilinear_interpolate(bev_feature[i], xs,
+                                                 ys)  # [5*num_box, C]
+
+            roi_feature = torch.cat([
+                features[i * num_boxes:(i + 1) * num_boxes]
+                for i in range(num_points)
+            ],
+                                    dim=1)  # [num_box, 5*C]
+            roi_features.append(roi_feature)
+
+        return roi_features
+
+    @staticmethod
+    def get_feature_points(bbox):
+        """Get feature points of bbox.
+
+        Args:
+            bbox (:obj:`BaseInstance3DBoxes`): Prediction bboxes
+
+        Returns:
+            torch.Tensor: The location of feature points. [5*num_box, 3]
+        """
+
+        # get corners
+        if isinstance(bbox, LiDARInstance3DBoxes):
+            front_left = bbox.corners[:, 4, :2].squeeze(1)  # [num_box, 2]
+            back_left = bbox.corners[:, 0, :2].squeeze(1)
+            front_right = bbox.corners[:, 7, :2].squeeze(1)
+            back_right = bbox.corners[:, 3, :2].squeeze(1)
+        elif isinstance(bbox, CameraInstance3DBoxes):
+            raise NotImplementedError
+        elif isinstance(bbox, DepthInstance3DBoxes):
+            raise NotImplementedError
+        else:
+            raise NotImplementedError
+
+        # get feature points
+        center = bbox.bottom_center[:, :2]  # [num_box, 2]
+        front = (front_left + front_right) / 2.0
+        back = (back_left + back_right) / 2.0
+        left = (front_left + back_left) / 2.0
+        right = (front_right + back_right) / 2.0
+        points = torch.cat([center, front, back, left, right],
+                           dim=0)  # [5*num_box, 2]
+
+        return points
+
+    @staticmethod
+    def bilinear_interpolate(bev_feature, x, y):
```

Comment:
Can this function be implemented with grid_sample?

Reply:
Nice advice. I reimplemented it with `grid_sample` and it's much faster than the original one. :smile: